numpyro/hmc_util.py

```diff
@@ -781,3 +782,112 @@ def single_chain_init(key, only_params=False):
+        joined_subposteriors = vmap(lambda x, idx: x[idx])(joined_subposteriors, draw_idxs)
+
+    if diagonal:
+        # compute weights for each subposterior (ref: Section 3.1 of [1])
+        weights = vmap(lambda x: (1 - 1 / n_samples) / np.var(x, axis=0))(joined_subposteriors)
+        normalized_weights = weights / np.sum(weights, axis=0)
+        # get weighted samples
+        samples_flat = np.einsum('ij,ikj->kj', normalized_weights, joined_subposteriors)
+    else:
+        weights = vmap(lambda x: np.linalg.inv(_cov(x)))(joined_subposteriors)
+        normalized_weights = np.matmul(np.linalg.inv(np.sum(weights, axis=0)), weights)
+        samples_flat = np.einsum('ijk,ilk->lj', normalized_weights, joined_subposteriors)
+
+    # unravel_fn acts on 1 sample of a subposterior
+    _, unravel_fn = ravel_pytree(tree_map(lambda x: x[0], subposteriors[0]))
+    return vmap(lambda x: unravel_fn(x))(samples_flat)
+
+
+def parametric(subposteriors, num_draws=None, diagonal=False, rng=None):
+    """
+    Merges subposteriors following (embarrassingly parallel) parametric Monte Carlo algorithm.
+
+    **References:**
+
+    1. *Asymptotically Exact, Embarrassingly Parallel MCMC*,
+       Willie Neiswanger, Chong Wang, Eric Xing
+
+    :param list subposteriors: a list in which each element is a collection of samples.
+    :param int num_draws: number of draws from the merged posterior.
+    :param jax.random.PRNGKey rng: source of the randomness, defaults to `jax.random.PRNGKey(0)`.
+    :return: if `num_draws` is None, returns the estimated mean and variance/covariance parameters
+        of the joined posterior; otherwise, returns a collection of `num_draws` samples with the
+        same data structure as each subposterior.
+    """
+    rng = random.PRNGKey(0) if rng is None else rng
+    joined_subposteriors = tree_multimap(lambda *args: np.stack(args), *subposteriors)
+    joined_subposteriors = vmap(vmap(lambda sample: ravel_pytree(sample)[0]))(joined_subposteriors)
+
+    submeans = np.mean(joined_subposteriors, axis=1)
+    n_samples = tree_flatten(subposteriors[0])[0][0].shape[0]
+    if diagonal:
+        # NB: jax.numpy.var does not support ddof=1, so we do it manually
+        weights = vmap(lambda x: (1 - 1 / n_samples) / np.var(x, axis=0))(joined_subposteriors)
+        var = 1 / np.sum(weights, axis=0)
+        normalized_weights = var * weights
+
+        # comparing to consensus implementation, we compute weighted mean here
+        mean = np.einsum('ij,ij->j', normalized_weights, submeans)
+    else:
+        weights = vmap(lambda x: np.linalg.inv(_cov(x)))(joined_subposteriors)
+        cov = np.linalg.inv(np.sum(weights, axis=0))
+        normalized_weights = np.matmul(cov, weights)
+
+        # comparing to consensus implementation, we compute weighted mean here
+        mean = np.einsum('ijk,ik->j', normalized_weights, submeans)
+
+    if num_draws is not None:
```

Comment:
nit: It will be nice if we can keep the function return type independent of the arguments. In this case, what do you think about refactoring so that `parametric` returned mean, cov, but we had another helper function called `parametric_draws` that calls parametric and returns `num_draws` samples? This latter function could be used in the tests then. This will avoid boiler-plate, but also keep function types consistent.